src/main/java/mate/academy/dao/BookDaoImpl.java

```diff
@@ -0,0 +1,156 @@
+            preparedStatement.setString(1, book.getTitle());
+            preparedStatement.setBigDecimal(2, book.getPrice());
+
+            int affectedRows = preparedStatement.executeUpdate();
+
+            try {
+                boolean b = affectedRows < 1;
+            } catch (DataProcessingException ex) {
+                throw new DataProcessingException("Expected to insert at least one row, "
+                        + "but inserted 0 rows.", ex);
+            }
+            ResultSet generatedKeys = preparedStatement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+                Long id = generatedKeys.getObject(1, Long.class);
+                book.setId(id);
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't create new book: " + book, e);
+        }
+        return book;
+    }
+
+    @Dao
+    @Override
+    public Optional<Book> findById(Long id) {
+        String sql = "SELECT * FROM books WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
+
+            preparedStatement.setLong(1, id);
+            ResultSet resultSet = preparedStatement.executeQuery();
+
+            if (resultSet.next()) {
+                Long rowId = resultSet.getObject("id", Long.class);
+                String title = resultSet.getString("title");
+                BigDecimal price = resultSet.getObject("price", BigDecimal.class);
+
+                Book book = new Book();
+                book.setId(rowId);
+                book.setTitle(title);
+                book.setPrice(price);
+                return Optional.of(book);
+            }
+
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't create connection", e);
+        }
+        return Optional.empty();
+    }
+
+    @Dao
+    @Override
+    public List<Book> findAll() {
+        Long id = 0L;
+        List<Book> bookList = new ArrayList<>();
+
+        String sql = "SELECT * FROM books WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
+
+            preparedStatement.setLong(1, id);
+            ResultSet resultSet;
+
+            do {
+                resultSet = preparedStatement.executeQuery();
+                Long rowId = resultSet.getObject("id", Long.class);
+                String title = resultSet.getString("title");
+                BigDecimal price = resultSet.getObject("price", BigDecimal.class);
+
+                Book book = new Book();
+                book.setId(rowId);
+                book.setTitle(title);
+                book.setPrice(price);
+
+                bookList.add(book);
+                id++;
+            } while (resultSet.next());

```

Comment:
The loop logic in `findAll` is flawed. You should not use a `do-while` loop with incrementing `id` to fetch all books. Instead, execute the query once and iterate over the `ResultSet` to add each book to the `bookList`.

Reply:
ok, done